Corrections, additions, test vectors (#1)
chris-wood/draft-bar-cfrg-spake2plus: draft-bar-cfrg-spake2plus.md

Brute-force resistance may be obtained through various computation hardness parameters such as memory or CPU cycles,
and are typically configurable.
Scrypt {{!RFC7914}} and Argon2 are common examples of PBKDF functions.
PBKDF and hardness parameter selection are out of scope of this document.
{{Ciphersuites}} specifies variants of KDF, MAC, and Hash
suitable for use with the protocols contained herein.

Let A and B be two parties. A and B may also have digital
representations of the parties' identities such as Media Access Control addresses
or other names (hostnames, usernames, etc). A and B may share additional data
(the context) separate from their identities which they may want to include in
the protocol transcript.
One example of additional data is a list of supported protocol versions if SPAKE2+ were
used in a higher-level protocol which negotiates the use of a particular PAKE. Another
example is the inclusion of PBKDF parameters and the application name.
Including those would ensure that both parties agree upon the same set of supported
protocols and use the same PBKDF parameters and therefore prevent downgrade and

> cross-protocol attacks.

Reviewer: ```suggestion
cross-protocol attacks. Specification of precise context values is out of scope for this document.
```

## SPAKE2+ {#spake2plus}

Let w0 and w1 be two integers derived by hashing the password pw with the
identities of the two participants, A and B. Specifically, compute
w0s || w1s = PBKDF(len(pw) || pw || len(A) || A || len(B) || B),
and then w0 = w0s mod p and w1 = w1s mod p.
If both identities A and B are absent, then w0s || w1s = PBKDF(pw), i.e.,
the length prefix is omitted as in {{setup}}.

> If both identities A and B are unknown at the time of deriving w0 and w1,

Reviewer: What if only one of them is unknown? Is only the unknown identity omitted?

Author: Yeah, that was the idea. Will clarify.

When executing SPAKE2+, A selects x uniformly at random from the
numbers in the range [0, p), and lets X=x\*P+w0\*M, then transmits pA=X to
B. Upon receipt of X, B computes h\*X and aborts if the result is equal
to I. B then selects y uniformly at random from the numbers in [0, p),
then computes Y=y\*P+w0\*N, and transmits pB=Y to A. Upon receipt of Y,
A computes h\*Y and aborts if the result is equal to I.

A computes Z as h\*x\*(Y-w0\*N), and V as h\*w1\*(Y-w0\*N). B computes Z as
h\*y\*(X-w0\*M) and V as h\*y\*L. Both share Z and V as common values.
It is essential that both Z and V be used in combination with the transcript to
derive the keying material. The protocol transcript encoding is shown below.

~~~
TT = len(Context) || Context ||
  || len(A) || A || len(B) || B
  || len(M) || M || len(N) || N
  || len(X) || X || len(Y) || Y
  || len(Z) || Z || len(V) || V
  || len(w0) || w0
~~~

Context is an application-specific customization string shared between both

> parties and SHOULD precede the remaining transcript. It might contain the

Reviewer: ```suggestion
parties and MUST precede the remaining transcript. It might contain the
```

name and version number of the higher-level protocol, or simply the name and version
number of the application. The context MAY include additional data such as the
chosen ciphersuite and PBKDF parameters like the iteration count or salt.
The context and its length prefix MAY be omitted.

If an identity is absent, it is omitted from the transcript entirely. For example,
if both A and B are absent, then TT = len(Context) || Context || len(M) || M || len(N) || N || len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.

> Likewise, if only A is absent, TT = len(Context) || Context || len(B) || B || len(M) || M || len(N) || N || len(X) || X || len(Y) || Y || len(Z) || Z || len(w0) || w0.

Reviewer: This is worrisome :-( The security proofs assume treat Hash as a RO, which is the case if the input is prefix-free, but how do we guarantee prefix-freedom if some values are optional? As an example, consider two different protocols Foo and Bar. In Foo, A's identity is omitted. In Bar, B's identity is omitted. Moreover, assume that there exists a party acting as B in Foo with identity (name) N and a party acting as A in Bar with identity (name) N'. If N = N', and if the context string is omitted from both protocols, we can have a RO collision. 

Is this a problem? (I think an easy fix is to add a type byte before each length value in the transcript, e.g.:

```
TT = 0x00 || len(Context) || Context ||
  0x01 || || len(A) || A || 0x02 || len(B) || B
  0x03 || || len(M) || M || 0x04 || len(N) || N
  0x05 || len(X) || X || 0x06 || len(Y) || Y
  0x07 || len(Z) || Z || 0x08 || len(V) || V
  0x09 || len(w0) || w0
```

Cumbersome, and probably a hard change to pitch, but I think this is probably needed.

Reviewer: (Alternatively, we can remove the "optional" part and just always encode the lengths. That'd probably be simpler.)

Author: That's a great point, thanks for bringing this up! I've removed the "optional" part and identities are now mandatory parts of the transcript, even if their lengths are zero and the identities themselves are empty octet strings. It even simplifies things a bit. 👍

This must only be done for applications in which identities are implicit. Otherwise,
the protocol risks Unknown Key Share attacks (discussion of Unknown Key Share attacks
in a specific protocol is given in {{?I-D.ietf-mmusic-sdp-uks}}.

Upon completion of this protocol, A and B compute shared secrets Ka, Ke, KcA,
and KcB as specified in {{keys}}. B MUST send A a key confirmation message cB
so both parties agree upon these shared secrets. This confirmation message cB
is computed as a MAC over the received share (pA) using KcB. Specifically, B
computes cB = MAC(KcB, pA), where MAC is also a secure PRF. After receipt and
verification of B's confirmation message, A MUST send B a confirmation message
using a MAC computed equivalently except with the use of pB and KcA. Key
confirmation verification requires recomputation of the MAC and checking

> for equality against that which was received.

Reviewer: How do we deal with tail loss? What if an adversary drops the confirmation message? It's probably better to say, B MUST NOT send application data to A until it has received and verified the confirmation message.

Author: Let's add that, good point.

# Key Schedule and Key Confirmation {#keys}

The protocol transcript TT, as defined in {{spake2plus}},
is unique and secret to A and B. Both parties use TT to
derive shared symmetric secrets Ke and Ka as Ke || Ka = Hash(TT). The length of each
key is equal to half of the digest output, e.g., |Ke| = |Ka| = 128 bits for SHA-256.
If the required key size is less than half the digest output, e.g. when using SHA-512
to derive two 128-bit keys, the digest output MAY be truncated.

Both endpoints use Ka to derive subsequent MAC keys for key confirmation messages.
Specifically, let KcA and KcB be the MAC keys used by A and B, respectively.
A and B compute them as KcA || KcB = KDF(nil, Ka, "ConfirmationKeys")

The length of each of KcA and KcB is equal to half of the KDF
output, e.g., |KcA| = |KcB| = 128 bits for HKDF-SHA256. If half of the KDF
output size exceeds the required key size for the chosen MAC, e.g. when using

> HKDF-SHA512 for CMAC-AES-128, the KDF output MAY be truncated.

Reviewer: ```suggestion
HKDF-SHA512 as the KDF and CMAC-AES-128 as the MAC, the KDF output MAY be truncated.
```

chosen ciphersuite and PBKDF parameters like the iteration count or salt.
The context and its length prefix MAY be omitted.

If an identity is absent, its length is given as zero and the identity itself
the empty octet string. If both A and B are absent, then both lengths are zero
and both A and B will be empty octet strings. In applications where identities
are not implicit, A and B SHOULD always be non-empty. Otherwise, the protocol
risks Unknown Key Share attacks (discussion of Unknown Key Share attacks in a

> specific protocol is given in {{?I-D.ietf-mmusic-sdp-uks}}.

Reviewer: ```suggestion
specific protocol is given in {{?I-D.ietf-mmusic-sdp-uks}}).
```

# Security Considerations

SPAKE2+ appears in {{TDH}} and is proven secure in {{UCAnalysis}}.

Beyond the cofactor multiplication checks to ensure that elements received from
a peer are in the prime order subgroup of G, they also MUST be checked for group

> membership as failure to properly validate group elements can lead to attacks.

Reviewer: Riad pointed out simpler text for this, captured in this issue: https://github.com/cfrg/draft-irtf-cfrg-hpke/issues/107

# Test Vectors {#testvectors}

This section contains test vectors for SPAKE2+ using

> the P256-SHA256-HKDF-HMAC/CMAC ciphersuite. (Choice of PBKDF is omitted

Reviewer: ```suggestion
the P256-SHA256-HKDF-HMAC and P256-SHA256-HKDF-CMAC ciphersuites. (Choice of PBKDF is omitted
```
